lib/pluginsync/github.rb

```diff
@@ -0,0 +1,208 @@
+        @repo = @gh.repo name
+        @owner = @repo.owner.login
+      end
+
+      def content(path, default=nil)
+        file = @gh.contents(@name, :path=>path)
+        Base64.decode64 file.content
+      rescue
+        nil
+      end
+
+      def upstream
+        if @repo.fork?
+          @repo.parent.full_name
+        else
+          nil
+        end
+      end
+
+      def ref_sha(ref, repo=@name)
+        refs = @gh.refs repo
+        if result = refs.find{ |r| r.ref == ref }
+          result.object.sha
+        else
+          nil
+        end
+      end
+
+      def sync_branch(branch, opt={})
+        parent = opt[:origin] || upstream || raise(ArgumentError, "Repo #{@name} is not a fork and no origin specified for syncing.")
+        origin_branch = opt[:branch] || 'master'
+
+        origin_sha = ref_sha("refs/heads/#{origin_branch}", parent)
+
+        fork_ref = "heads/#{branch}"
+        fork_sha = ref_sha("refs/heads/#{branch}")
+
+        if ! fork_sha
+          @gh.create_ref(@name, fork_ref, origin_sha)
+        elsif origin_sha != fork_sha
+          begin
+            @gh.update_ref(@name, fork_ref, origin_sha)
+          rescue Octokit::UnprocessableEntity
+            @log.warn "Fork #{name} is out of sync with #{parent}, syncing to #{name} #{origin_branch}"
+            origin_sha = ref_sha("refs/heads/#{origin_branch}")
+            @gh.update_ref(@name, fork_ref, origin_sha)
+          end
+        end
+      end
+
+      def update_content(path, content, opt={})
+        branch = opt[:branch] || "master"
+
+        raise(Argument::Error, "This tool cannot directly commit to #{INTEL_ORG} repos") if @name =~ /^#{INTEL_ORG}/
+        raise(Argument::Error, "This tool cannot directly commit to master branch") if branch == 'master'
+
+        message = "update #{path} by pluginsync tool"
+        content = Base64.encode64 content
+
+        ref = "heads/#{branch}"
+        latest_commit = @gh.ref(@name, ref).object.sha
+        base_tree = @gh.commit(@name, latest_commit).commit.tree.sha
+
+        sha = @gh.create_blob(@name, content, "base64")
+        new_tree = @gh.create_tree(
+          @name,
+          [ {
+            :path => path,
+            :mode => "100644",
+            :type => "blob",
+            :sha => sha
+          } ],
+          { :base_tree => base_tree }
+        ).sha
+
+        new_commit = @gh.create_commit(@name, message, new_tree, latest_commit).sha
+        @gh.update_ref(@name, ref, new_commit) if branch
+      end
+
+      def create_pull_request(branch, message)
+        @gh.create_pull_request(upstream, "master", "#{@repo.owner.login}:#{branch}", message)
+      end
+
+      def yml_content(path, default={})
+        YAML.load(content(path))
+      rescue
+        default
+      end
+
+      def plugin_name
+        @name.match(/snap-plugin-(collector|processor|publisher)-(.*)$/)
```

Comment:
@kindermoumoute nope - kubesnap project will merge into snap plugins in the longer run.